### ISSUE TITLE
Trainer hooks for before training start and after test/val ends

### ISSUE THREAD

**Issue by chnsh** [PyTorchLightning/pytorch-lightning#1664]
#### What is your question?

I am wondering if there are existing hooks that allow me to do something before training starts and after test ends.

Use case:
1. **Training start**: Initialization works like call self.logger.watch(self) - right now I am using prepare_data as a hack since I could not find something like this.

2. **On test end/val end**: My research requires me to build some plots with predictions (and ground truth) of the learned model and I need all datapoints to do some aggregated processing so I cannot just use a current batch and I cannot use the `epoch_end` hooks since I really want to do this as a last step (end of last epoch). I used to do that as a last step in ignite (since then I have moved  to lightning and love it)

EDIT: On a second thought, if I call test after fit has completed, then `test_epoch_end()` will work since it will be called only once?

EDIT2: Is there a place in the documentation where I can see a comprehensive list of callbacks and hooks and the order in which everything is called? I am a little confused - I see hooks and callbacks and and I feel confused about when to use which, is my ask better suited to be used with callbacks and I should build one myself?

**Comment by olineumann** [comment]
As mentioned in the docs Callbacks should be used for not essential logic. Init a logger with self.logger.watch would be necessary so use hooks. But why you have to init the logger like this? 

The second case: Plot some images after each epoch would be no essential logic so should be implemented with Callbacks.

I hope it helps. Don't know if it's necessary to define a hook for training_start. It could be discussed here.

[Callback Doc](https://pytorch-lightning.readthedocs.io/en/latest/callbacks.html)

**Comment by chnsh** [comment]
Thanks for replying, the logger that I use (WandB) has a feature that allows it to monitor gradients and parameters if I call `watch`.

The other question I have is with respect to the docs, I see a method `on_train_start` but do not find it in the list of hooks above, is there a reason for that?

**Comment by chnsh** [comment]
Also, for the after test ends logic - are there examples of how to use callbacks to access predictions?

I do not have to log everything after every test epoch, I want to collect all predictions that the test step has access to and then get them together and aggregate them for my use.

Is there a way to easily access that in a callback?

**Comment by olineumann** [comment]
> Thanks for replying, the logger that I use (WandB) has a feature that allows it to monitor gradients and parameters if I call `watch`.
> 
> The other question I have is with respect to the docs, I see a method `on_train_start` but do not find it in the list of hooks above, is there a reason for that?

Would it work if you call watch after init pl_module but before start training? But you could use Callbacks or Hooks for that also. Notice: There is a watch method for the wandb lightning logger which just calls the wandb.watch() method.

Wouldn't this work:
```python
model = Some_Lightning_Module
wandb_logger = WandbLogger(...)
wandb_logger.watch(model)
trainer = Trainer()
trainer.fit()
```

[CODE](https://github.com/PyTorchLightning/PyTorch-Lightning/blob/0.6.0/pytorch_lightning/logging/wandb.py#L74-L75)

The second question: There is an 'on_training_start' hook and callback. You should be able to use both.
[Hook](https://github.com/PyTorchLightning/PyTorch-Lightning/blob/master/pytorch_lightning/core/hooks.py#L26-L29)
[Callback](https://github.com/PyTorchLightning/PyTorch-Lightning/blob/master/pytorch_lightning/callbacks/base.py#L65-L67)

> Also, for the after test ends logic - are there examples of how to use callbacks to access predictions?
> 
> I do not have to log everything after every test epoch, I want to collect all predictions that the test step has access to and then get them together and aggregate them for my use.
> 
> Is there a way to easily access that in a callback?

What do you mean by accessing predictions? The trainer should get only loss or other log values from the lightning module. So to get predictions, you should save it e.g. in the module itself. Correct me here if I'm wrong. My first idea would be to use the test functionality. So implement a test in the lightning module and call the wandb logger by self.logger. After training or after several epochs, you can call model.test() which would run a test and log e.g. images to wandb. That would be my first approach. Maybe there is a better way.

**Comment by chnsh** [comment]
@olineumann both are great approaches - while waiting for your reply, that is what I ended up doing.

From my test method, I accumulate predictions in a tensor and then access it using `pl_module.predictions()` in the callback. 

This works well. Originally I wasn't sure (I still am not) if this was the correct way because this is "leaking abstraction" from the module to the callback - this code will only work with my particular pl module and in general is not a great practice (but works for now - so feel free to close my the issue)

**Comment by olineumann** [comment]
@chnsh I'm always happy to help :)

What do you mean by accumulating? Simply gather predictions in an array or make some sort of reduction like calculating the average image or similar. When I said I would use the test functionality, I would log the results also in this method. This is only possible if every image of the test set is logged and not an average or some other accumulation.

I think your way is absolutely okay. Your model saves predictions in an attribute and other objects can access it via predictions(). In the end you pl_module ist a fully functional pl_module with some decorations on it. So every kind of trainer, logger, checkpoint, callback, ... can be used. So in my opinion, it is also correct from a software architectural view. Maybe I would rename the predictions method to get_test_predicitons() or so to avoid collisions and make it more clear that it is only a get function for test predictions and not a function which makes predictions. 

If you want you could also pass the current prediction to the callback by adding it to the dict.
```python
    def validation_step(self, batch, batch_idx):
        ...
        images.append(prediction)
        ...
        if batch % 100 == 0: # if do not want to log every prediction
            return {'val_loss': getattr(F, self.hparams.loss)(y_hat, y), 'images': images}
        else:
            return {'val_loss': getattr(F, self.hparams.loss)(y_hat, y)}
```
and then accessing it in the callback
```python
  def on_epoch_end(self, trainer, pl_module):
        logs = trainer.callback_metrics
        if 'images' in logs:
            # do stuff on or with images
            pl_module.logger.log_image(logs['images']) # check if logger is wandb or at least contains one
```

Not sure how the trainer would accumulate the images... But this could be overwritten in some way. In my opinion, you can leave as you do it currently.

I do not have the right to close this issue but you can.

**Comment by chnsh** [comment]
Thanks for replying @olineumann !

What I mean by leaking abstraction is that the method `get_test_predictions()` is not in the regular pl_module and if I were to use a different pl_module that does implement this method, the code would break which is why it felt like the incorrect thing to do.

Your second method looks interesting though, I can accumulate these predictions/images in an array and use callback_metrics dict to access the tensor - that sounds correct and does not break the abstraction, right?

**Comment by olineumann** [comment]
> Thanks for replying @olineumann !
> 
> What I mean by leaking abstraction is that the method `get_test_predictions()` is not in the regular pl_module and if I were to use a different pl_module that does implement this method, the code would break which is why it felt like the incorrect thing to do.

Only if you use the same Callback class which assumes that there is such a method and only handles image logging to wandb. So when you change the module you would either implement the method or leave the callback object. But yes, it would be more elegant if you can change both without care. 

> 
> Your second method looks interesting though, I can accumulate these predictions/images in an array and use callback_metrics dict to access the tensor - that sounds correct and does not break the abstraction, right?

Yes when I'm right, the Trainer simply collects the dict elements and needs at least 'loss' returned. Only the logger needs a function to aggregate the elements to (I think only) one value. See: agg_key_funcs in [LoggerBase](https://pytorch-lightning.readthedocs.io/en/0.7.4/pytorch_lightning.loggers.base.html)
